File: 3DGame/snake.py
import pygame
from pygame.locals import *
from OpenGL.GL import *
from OpenGL.GLUT import *
from OpenGL.GLU import *
import random

# Constants for the game
WIDTH, HEIGHT, DEPTH = 10, 10, 10
BLOCK_SIZE = 1
SPEED = 0.2

# Directions for the snake
DIRECTIONS = {
    "UP": (0, 1, 0),
    "DOWN": (0, -1, 0),
    "LEFT": (-1, 0, 0),
    "RIGHT": (1, 0, 0),
    "FORWARD": (0, 0, 1),
    "BACKWARD": (0, 0, -1)
}

class SnakeGame3D:
    def __init__(self):
        self.snake = [(0, 0, 0)]  # Snake starts with one block
        self.direction = DIRECTIONS["UP"]
        self.food = self.place_food()
        self.score = 0
        self.game_started = False  # Flag to check if the game has started

    def place_food(self):
        while True:
            food = (random.randint(-WIDTH // 2, WIDTH // 2),
                    random.randint(-HEIGHT // 2, HEIGHT // 2),
                    random.randint(-DEPTH // 2, DEPTH // 2))
            if food not in self.snake:
                return food

    def move_snake(self):
        head = self.snake[-1]
        new_head = (head[0] + self.direction[0],
                    head[1] + self.direction[1],
                    head[2] + self.direction[2])
        
        if (new_head in self.snake or
            not (-WIDTH // 2 <= new_head[0] <= WIDTH // 2) or
            not (-HEIGHT // 2 <= new_head[1] <= HEIGHT // 2) or
            not (-DEPTH // 2 <= new_head[2] <= DEPTH // 2)):
            return False  # Game over

        self.snake.append(new_head)
        if new_head == self.food:
            self.food = self.place_food()
            self.score += 1
        else:
            self.snake.pop(0)
        
        return True

    def change_direction(self, new_direction):
        # Prevent reversing direction
        opposite = tuple(-x for x in self.direction)
        if new_direction != opposite:
            self.direction = new_direction
            self.game_started = True  # Start the game when the direction changes

    def draw_block(self, position, color):
        x, y, z = position
        glColor3f(*color)
        glPushMatrix()
        glTranslatef(x, y, z)
        glutSolidCube(BLOCK_SIZE)
        glPopMatrix()

    def render(self):
        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)
        glLoadIdentity()

        # Move camera back to view the game space
        glTranslatef(0.0, 0.0, -20)

        # Draw the snake
        for segment in self.snake:
            self.draw_block(segment, (0, 1, 0))

        # Draw the food
        self.draw_block(self.food, (1, 0, 0))

        pygame.display.flip()

def main():
    # Initialize GLUT
    glutInit()

    pygame.init()
    display = (800, 600)
    pygame.display.set_mode(display, DOUBLEBUF | OPENGL)
    gluPerspective(45, (display[0] / display[1]), 0.1, 50.0)

    # Enable depth testing
    glEnable(GL_DEPTH_TEST)

    game = SnakeGame3D()
    clock = pygame.time.Clock()

    running = True
    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
            elif event.type == pygame.KEYDOWN:
                if event.key == pygame.K_w:
                    game.change_direction(DIRECTIONS["UP"])
                elif event.key == pygame.K_s:
                    game.change_direction(DIRECTIONS["DOWN"])
                elif event.key == pygame.K_a:
                    game.change_direction(DIRECTIONS["LEFT"])
                elif event.key == pygame.K_d:
                    game.change_direction(DIRECTIONS["RIGHT"])
                elif event.key == pygame.K_q:
                    game.change_direction(DIRECTIONS["FORWARD"])
                elif event.key == pygame.K_e:
                    game.change_direction(DIRECTIONS["BACKWARD"])

        if game.game_started:  # Only move the snake once the game starts
            if not game.move_snake():
                print(f"Game Over! Your score: {game.score}")
                running = False

        game.render()
        clock.tick(10)  # Control game speed

    pygame.quit()

if __name__ == "__main__":
    main()

File: 3DGame/snake.js
// Constants for the game
const WIDTH = 20, HEIGHT = 20, DEPTH = 20; // Larger play area
const BLOCK_SIZE = 1;
const SPEED = 0.2;

// Directions for the snake
const DIRECTIONS = {
  "UP": [0, 1, 0],
  "DOWN": [0, -1, 0],
  "LEFT": [-1, 0, 0],
  "RIGHT": [1, 0, 0],
  "FORWARD": [0, 0, 1],
  "BACKWARD": [0, 0, -1]
};

let snake = [[0, 0, 0]]; // Snake starts with one block
let direction = DIRECTIONS["UP"];
let food = placeFood();
let score = 0;

function placeFood() {
  while (true) {
    let food = [
      Math.floor(Math.random() * (WIDTH + 1)) - WIDTH / 2,
      Math.floor(Math.random() * (HEIGHT + 1)) - HEIGHT / 2,
      Math.floor(Math.random() * (DEPTH + 1)) - DEPTH / 2
    ];
    // Ensure food is not placed on the snake
    if (!snake.some(segment => segment[0] === food[0] && segment[1] === food[1] && segment[2] === food[2])) {
      return food;
    }
  }
}

function moveSnake() {
  let head = snake[snake.length - 1];
  let newHead = [
    head[0] + direction[0],
    head[1] + direction[1],
    head[2] + direction[2]
  ];

  // Check if snake hits the walls or itself
  if (
    newHead[0] < -WIDTH / 2 || newHead[0] > WIDTH / 2 ||
    newHead[1] < -HEIGHT / 2 || newHead[1] > HEIGHT / 2 ||
    newHead[2] < -DEPTH / 2 || newHead[2] > DEPTH / 2 ||
    snake.some(segment => segment[0] === newHead[0] && segment[1] === newHead[1] && segment[2] === newHead[2])
  ) {
    return false; // Game over
  }

  snake.push(newHead);
  if (newHead[0] === food[0] && newHead[1] === food[1] && newHead[2] === food[2]) {
    food = placeFood(); // Place new food
    score++;
  } else {
    snake.shift(); // Remove last segment of snake
  }

  return true;
}

function changeDirection(newDirection) {
  // Prevent reversing direction
  let opposite = direction.map(x => -x);
  if (JSON.stringify(newDirection) !== JSON.stringify(opposite)) {
    direction = newDirection;
  }
}

function drawBlock(position, color) {
  const [x, y, z] = position;
  glColor3f(...color);
  glPushMatrix();
  glTranslatef(x, y, z);
  glutSolidCube(BLOCK_SIZE);
  glPopMatrix();
}

function drawWalls() {
  const wallColor = [0.5, 0.5, 0.5]; // Gray color for walls

  // Draw walls along X, Y, and Z axis at boundaries
  const wallThickness = 0.5;

  // X walls (front and back)
  for (let z = -DEPTH / 2 - 1; z <= DEPTH / 2 + 1; z++) {
    drawBlock([-WIDTH / 2 - 1, 0, z], wallColor); // Left wall
    drawBlock([WIDTH / 2 + 1, 0, z], wallColor); // Right wall
  }

  // Y walls (left and right)
  for (let z = -DEPTH / 2 - 1; z <= DEPTH / 2 + 1; z++) {
    drawBlock([0, -HEIGHT / 2 - 1, z], wallColor); // Bottom wall
    drawBlock([0, HEIGHT / 2 + 1, z], wallColor); // Top wall
  }

  // Z walls (near and far)
  for (let y = -HEIGHT / 2 - 1; y <= HEIGHT / 2 + 1; y++) {
    drawBlock([0, y, -DEPTH / 2 - 1], wallColor); // Near wall
    drawBlock([0, y, DEPTH / 2 + 1], wallColor); // Far wall
  }
}

function render() {
  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
  glLoadIdentity();

  // Move camera back
  glTranslatef(0.0, 0.0, -30);

  // Draw the walls
  drawWalls();

  // Draw the snake
  for (let segment of snake) {
    drawBlock(segment, [0, 1, 0]); // Snake color
  }

  // Draw the food
  drawBlock(food, [1, 0, 0]); // Food color

  // Display the score
  document.getElementById("score").innerText = Score: ${score};

  // Flip the display
  pygame.display.flip();
}

// Main game loop
function gameLoop() {
  if (!moveSnake()) {
    alert(Game Over! Your score: ${score});
    resetGame();
  }

  render();
  requestAnimationFrame(gameLoop); // Keep the game running
}

function resetGame() {
  snake = [[0, 0, 0]];
  direction = DIRECTIONS["UP"];
  food = placeFood();
  score = 0;
}

document.addEventListener("keydown", (event) => {
  switch (event.key) {
    case "w":
      changeDirection(DIRECTIONS["UP"]);
      break;
    case "s":
      changeDirection(DIRECTIONS["DOWN"]);
      break;
    case "a":
      changeDirection(DIRECTIONS["LEFT"]);
      break;
    case "d":
      changeDirection(DIRECTIONS["RIGHT"]);
      break;
    case "q":
      changeDirection(DIRECTIONS["FORWARD"]);
      break;
    case "e":
      changeDirection(DIRECTIONS["BACKWARD"]);
      break;
  }
});

function startGame() {
  // Initialize Pygame and OpenGL
  pygame.init();
  let display = [1000, 800];
  pygame.display.set_mode(display, DOUBLEBUF | OPENGL);
  gluPerspective(45, display[0] / display[1], 0.1, 50.0);

  glEnable(GL_DEPTH_TEST);

  gameLoop(); // Start the game loop
}

// Wait until the page is fully loaded
window.onload = startGame;
